src/udp_engine.cpp

```diff
@@ -199,6 +199,12 @@ void zmq::udp_engine_t::plug (io_thread_t *io_thread_, session_base_t *session_)
         errno_assert (rc == 0);
 #endif
 
+#ifdef SO_REUSEPORT
```

Comment:
Is this intentionally done for both unicast and multicast addresses? #3236 was referring specifically to multicast addresses.

Reply:
I've left this in for both address types as I was having trouble with both multicast and wildcard addresses. This solved my issue when using a non-wildcard address (in this case an IP in the multicast address range) but still caused an issue when using a wildcard address like udp://*:port but I think it might be an unrelated issue. 

My apologies for the multicast confusion, I can try using a fixed IP instead of one from the 239.*.*.* and see if that affects anything.

Reply:
I think if you set SO_REUSEADDR on multiple unicast sockets the data arriving on the system will be load balanced between them.  That is why the linux limits processes using SO_REUSEADDR to the same user ID.

Reply:
The behaviour depends on the OS (in particular between Linux and BSD systems), and somewhat differs between multicast and unicast addresses. https://stackoverflow.com/a/14388707/2019765 has quite comprehensive information on this.

Reply:
@sigiesec I've written a [gist](https://gist.github.com/Mystfit/6c015257b637ae31bcb63130da67627c) to test this. Multicast works as expected but using udp://127.0.0.1:40000 results in the first bound dish to hog all messages, even though a second program is bound to the same address:port. Killing the first bound program lets the second waiting program receive the messages instead.